pallets/liquidity-pools-forwarder/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		},
+	}
+
+	/// Maps a router id to its forwarding info.
+	///
+	/// Can only be mutated via admin origin.
+	#[pallet::storage]
+	pub type RouterForwarding<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::RouterId, ForwardInfo, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The router id does not have any forwarder info stored
+		ForwardInfoNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(0)]
+		pub fn set_forwarder(
+			origin: OriginFor<T>,
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::insert(
+				&router_id,
+				ForwardInfo {
+					source_domain,
+					contract: forwarding_contract,
+				},
+			);
+
+			Self::deposit_event(Event::<T>::ForwarderSet {
+				router_id,
+				source_domain,
+				forwarding_contract,
+			});
+
+			Ok(())
+		}
+
+		/// Remove the forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(1)]
+		pub fn remove_forwarder(origin: OriginFor<T>, router_id: T::RouterId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			if let Some(info) = RouterForwarding::<T>::take(&router_id) {
+				Self::deposit_event(Event::<T>::ForwarderRemoved {
+					router_id,
+					source_domain: info.source_domain,
+					forwarding_contract: info.contract,
+				});
+			}
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> MessageSender for Pallet<T> {
+		type Message = T::Message;
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn send(
+			router_id: T::RouterId,
+			origin: DomainAddress,
+			message: T::Message,
+		) -> DispatchResult {
+			let payload = if let Some(info) = RouterForwarding::<T>::get(&router_id) {
+				let wrapped =
+					T::Message::try_wrap_forward(info.source_domain, info.contract, message)?;
+				wrapped.serialize()
+			} else {
+				message.serialize()
+			};
+
+			T::MessageSender::send(router_id, origin, payload)
+		}
+	}
+
+	impl<T: Config> MessageReceiver for Pallet<T> {
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn receive(
+			router_id: T::RouterId,
+			domain_address: DomainAddress,
+			payload: Vec<u8>,
+		) -> DispatchResult {
+			let message = T::Message::deserialize(&payload)?;
+
+			// If message can be unwrapped, it was forwarded
+			if let Some((source_domain, _, lp_message)) = message.unwrap_forwarded() {
+				let router_ids = T::RouterProvider::routers_for_domain(source_domain);
+				for router_id in router_ids {
```

Comment:
Nit - use `router_ids.iter().find()`.

Reply:
I think we "cannot" due to the return

Reply:
Hmm, on second though, I think we can drop this whole loop. Basically, we want to get the `ForwardingInfo` for the provided router_id, if any, and then continue processing?

Reply:
Ouch, I was completely ignoring the input argument `router_id`. You are right @cdamian !

Reply:
Also, would the following work?

```
match message.unwrap_forwarded() {
				Some((source_domain, _, lp_message)) => {
					let router_ids = T::RouterProvider::routers_for_domain(source_domain);

					ensure!(
						router_ids.iter().any(|x| x == &router_id),
						DispatchError::Other("router ID not found for source domain")
					);

					match RouterForwarding::<T>::get(&router_id) {
						Some(_) => T::MessageReceiver::receive(
							router_id,
							domain_address,
							lp_message.serialize(),
						),
						None => Err(Error::<T>::ForwardInfoNotFound.into()),
					}
				}
				None => T::MessageReceiver::receive(router_id, domain_address, payload),
			}
```

Mainly too see if I got this flow right.

Reply:
Check my solution here: [ce64041](https://github.com/centrifuge/centrifuge-chain/pull/1971/commits/ce64041e4125bca42a8626418dccaebf663ce9a5)